crates/oxc_linter/src/rules/no_array_constructor.rs

```diff
@@ -0,0 +1,97 @@
+use oxc_ast::{
+    ast::{CallExpression, Expression, NewExpression},
+    AstKind, Span,
+};
+use oxc_diagnostics::{
+    miette::{self, Diagnostic},
+    thiserror::Error,
+};
+
+use crate::{context::LintContext, rule::Rule, AstNode};
+
+#[derive(Debug, Error, Diagnostic)]
+#[error("eslint(no-array-constructor): Disallow `Array` constructors")]
+#[diagnostic()]
+struct NoArrayConstructorDiagnostic(#[label] pub Span);
+
+#[derive(Debug, Default, Clone)]
+pub struct NoArrayConstructor;
+
+const RULE_NAME: &str = "no-array-constructor";
+
+impl Rule for NoArrayConstructor {
+    const NAME: &'static str = RULE_NAME;
+
+    fn run<'a>(&self, node: &AstNode<'a>, ctx: &LintContext<'a>) {
+        match node.get().kind() {
+            AstKind::CallExpression(CallExpression {
+                span,
+                arguments,
+                callee: Expression::Identifier(ident),
+                type_parameters,
+                optional,
+                ..
+            }) if arguments.len() != 1
+                && ident.name == "Array"
+                && type_parameters.is_none()
+                && !optional =>
+            {
+                ctx.diagnostic(NoArrayConstructorDiagnostic(*span));
+            }
+            AstKind::NewExpression(NewExpression {
+                span,
+                arguments,
+                callee: Expression::Identifier(ident),
+                type_parameters,
+                ..
+            }) if arguments.len() != 1 && ident.name == "Array" && type_parameters.is_none() => {
+                ctx.diagnostic(NoArrayConstructorDiagnostic(*span));
+            }
+            _ => {}
```

Comment:
Which style do you prefer?

```rust
        let (node, callee, arguments, type_parameters, optional) = match node.get().kind() {
            AstKind::CallExpression(call_expr) => (
                call_expr.node,
                &call_expr.callee,
                &call_expr.arguments,
                &call_expr.type_parameters,
                call_expr.optional,
            ),
            AstKind::NewExpression(new_expr) => (
                new_expr.node,
                &new_expr.callee,
                &new_expr.arguments,
                &new_expr.type_parameters,
                false,
            ),
            _ => return,
        };

        if let Expression::Identifier(ident) = callee
            && ident.name == "Array"
                && arguments.len() != 1
                && type_parameters.is_none()
                && !optional
        {
            ctx.diagnostic(NoArrayConstructorDiagnostic(*span));
            return

        }
```

Reply:
I don't really like the destructing pattern because it's hard to follow where the if-conditions are.

Reply:
Definitely looks much better!

Is this using `if-let-chains`?

Is it here?
```rust
if let Expression::Identifier(ident) = callee
    && ident.name == "Array"
    && arguments.len() != 1
    && type_parameters.is_none()
    && !optional
{
    // ...
}
```

Reply:
yes this is, rustc will warn you if you remove the feature flag from `lib.rs`